File: go/leetcode/31.下一个排列.go
/*
 * @lc app=leetcode.cn id=31 lang=golang
 *
 * [31] 下一个排列
 */

func swap(nums []int, i int, j int) {
	tmp := nums[j]
	nums[j] = nums[i]
	nums[i] = tmp
}

func reverse(nums []int, i int, j int)  {
	for i < j {
		swap(nums, i, j)
		i++
		j--
	}
}

func nextPermutation(nums []int)  {
	// 1 2 4 7 5 3 1 
	// 1 2 5 7 4 3 1
	// 1 2 5 1 3 4 7
	length := len(nums)
	if length < 2 {
		return 
	}
	var i, j int
	for i = length - 2; i >= 0; i-- {
		if nums[i] < nums[i+1] {
			for j = length-1; j > i; j-- {
				if nums[j] > nums[i] {
					break
				}
			}
			swap(nums, i, j)
			reverse(nums, i+1, length-1)
			break
		}
	}
	if i < 0 {
		reverse(nums, 0, length-1)
	}
}

File: cpp/leetcode/31.下一个排列.cpp
/*
 * @lc app=leetcode.cn id=31 lang=cpp
 *
 * [31] 下一个排列
 *
 * https://leetcode-cn.com/problems/next-permutation/description/
 *
 * algorithms
 * Medium (37.10%)
 * Likes:    1237
 * Dislikes: 0
 * Total Accepted:    188.9K
 * Total Submissions: 507.5K
 * Testcase Example:  '[1,2,3]'
 *
 * 实现获取 下一个排列
 * 的函数，算法需要将给定数字序列重新排列成字典序中下一个更大的排列。
 *
 * 如果不存在下一个更大的排列，则将数字重新排列成最小的排列（即升序排列）。
 *
 * 必须 原地 修改，只允许使用额外常数空间。
 *
 *
 *
 * 示例 1：
 *
 *
 * 输入：nums = [1,2,3]
 * 输出：[1,3,2]
 *
 *
 * 示例 2：
 *
 *
 * 输入：nums = [3,2,1]
 * 输出：[1,2,3]
 *
 *
 * 示例 3：
 *
 *
 * 输入：nums = [1,1,5]
 * 输出：[1,5,1]
 *
 *
 * 示例 4：
 *
 *
 * 输入：nums = [1]
 * 输出：[1]
 *
 *
 *
 *
 * 提示：
 *
 *
 * 1
 * 0
 *
 *
 */

#include <algorithm>
#include <vector>
// @lc code=start
class Solution {
public:
  void nextPermutation(std::vector<int> &nums) {
    // [1,4,9,8,3,2] => [1,8,2,3,4,9] => swap48，revert9432
    if (nums.size() <= 1) {
      return;
    }
    int i = nums.size() - 2;
    while (i >= 0 && nums[i] >= nums[i + 1]) {
      i--;
    }
    if (i >= 0) { // found
      int j = nums.size() - 1;
      while (nums[j] <= nums[i]) {
        j--;
      }
      std::swap(nums[i], nums[j]);
    }
    std::reverse(nums.begin() + i + 1, nums.end());
  }
};
// @lc code=end
